onnxruntime/core/providers/webnn/builders/impl/pad_op_builder.cc

```diff
@@ -116,11 +121,42 @@ Status PadOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
     options.set("value", value);
   }
 
+  // Padding of WebNN cannot be negative.
+  std::vector<int32_t> webnn_start_padding;
+  std::vector<int32_t> webnn_end_padding;
+  bool neg_padding = false;
+  if (hasNegativeValue(start_padding)) {
+    neg_padding = true;
+    std::transform(start_padding.begin(), start_padding.end(), std::back_inserter(webnn_start_padding),
+                   [](auto x) { return std::max(x, 0); });
+  } else {
+    webnn_start_padding = start_padding;
+  }
+
+  if (hasNegativeValue(end_padding)) {
+    neg_padding = true;
+    std::transform(end_padding.begin(), end_padding.end(), std::back_inserter(webnn_end_padding),
+                   [](auto x) { return std::max(x, 0); });
+  } else {
+    webnn_end_padding = end_padding;
+  }
+
   emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("pad", input,
-                                                                            emscripten::val::array(start_padding),
-                                                                            emscripten::val::array(end_padding),
+                                                                            emscripten::val::array(webnn_start_padding),
+                                                                            emscripten::val::array(webnn_end_padding),
                                                                             options);
-
+  // Handle the negative padding with slice.
+  if (neg_padding) {
+    std::vector<int32_t> start;
+    std::vector<int32_t> size;
+    for (size_t i = 0; i < start_padding.size(); i++) {
+      start.push_back(start_padding[i] >= 0 ? 0 : -start_padding[i]);
+      size.push_back(SafeInt<int32_t>(input_shape[i] + start_padding[i] + end_padding[i]));
```

Comment:
```suggestion
      size.push_back(SafeInt<int32_t>(input_shape[i]) + start_padding[i] + end_padding[i]);
```

The safeint has to wrap the numeric type *before* applying the math operation. Otherwise overflow has already occurred before reaching the SafeInt constructor.

Reply:
Calculating int64+uint32+uint32 will be operated in int64 for integer promotion, while uint32+uint32+uint32 will be operated in uint32. If overflow occurs in int64, it will also occur in uint32. However, overflow in uint32 does not necessarily occur in int64, e.g. UINT_MAX+1+1. So I guess that applying safeint after applying the math operation is safer? In some case, it is only valid after applying math operation like SafeInt<int32_t>((INT_MAX+4LL)/10).

BTW, I find that CPU EP  tends to use before math operation, while other EPs prefer after math operations, e.g.  [nnapi unsqueeze](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/unsqueeze_op_builder.cc#L60).

Reply:
Ah, you are correct - `input_shape` is `int64`. I thought `input_shape` was int32.

> I find that CPU EP tends to use before math operation, while other EPs prefer after math operations
It's *generally* more robust to wrap the SafeInt early, so that all subsequent operations are validated, rather than wrap with SafeInt at the final cast. (but it is fine in this case above, either way)